iota_client_get_inputs enhancement

howjmay opened the pull request:
Generate one new address and fetch one address to `iota_client_get_balances`
each time. And end this function `iota_client_get_inputs` once it meets
the threshold (minimun needed balances)

The time cost of  `iota_client_get_balances` and `iota_client_get_new_address` are bascially the same on my laptop, so I choosed to wrap them in the same loop.
close #1199

oopsmonk commented:
@HowJMay thanks, can you explant how can this modification will fix #1199? I don't see any changes in the result.

howjmay commented:
> @HowJMay thanks, can you explant how can this modification will fix #1199? I don't see any changes in the result.

I might misunderstand the description of #1199.
I break the loop once the `out_inputs->total_balance` large or equal to minimun required balance. Thus, the loop doesn't need to fetch all the balances in the blanaces list.

And the if statement below can just check out the bool variable `balance_sufficient` to determine if the seed has sufficient balances for transaction

oopsmonk commented:
> > @HowJMay thanks, can you explant how can this modification will fix #1199? I don't see any changes in the result.
> 
> I might misunderstand the description of #1199.
> I break the loop once the `out_inputs->total_balance` large or equal to minimun required balance. Thus, the loop doesn't need to fetch all the balances in the blanaces list.
> 
> And the if statement below can just check out the bool variable `balance_sufficient` to determine if the seed has sufficient balances for transaction

Okay, the problem is not at the balance checking. 
The problem is `balances_req` containing a list of addresses when calling `iota_client_get_balances` so the API will not stop before fetching all balances in the list.

howjmay commented:
Thank you for your explanation. May I close this PR, and reopen it once I finish it in a right way?

oopsmonk commented:
> Thank you for your explanation. May I close this PR, and reopen it once I finish it in a right way?

sure, thank for your help :-)

howjmay commented:
@oopsmonk I plan to do in the following way
I will fetch 5 addresses each time I call `iota_client_get_balances`
but I need to new/free balances_[req, res] object each time before I call `iota_client_get_balances`.
I am not sure is it a acceptable way to solve this issue?

oopsmonk commented:
> @oopsmonk I plan to do in the following way
> I will fetch 5 addresses each time I call `iota_client_get_balances`
> but I need to new/free balances_[req, res] object each time before I call `iota_client_get_balances`.
> I am not sure is it a acceptable way to solve this issue?

you can benchmark your solution, then you can figure out it.

howjmay committed:
iota_client_get_inputs enhancement

Generate one new address and fetch one address to `iota_client_get_balances`
each time. And terminate this function `iota_client_get_inputs` once it meets
the threshold (minimun needed balances)

For requests which don't assign addr_opt.total, the if statement in the loop
would examine is the current address is an unused address. If the current
address is an unused address, then break this loop.

howjmay commented:
@oopsmonk  I have pushed new commit. Is it a more appropriate solution?

oopsmonk commented on cclient/api/extended/get_inputs.c:
```diff
@@ -13,51 +13,60 @@
-    goto done;
-  }
+  address_opt_t addr_opt_elt = {addr_opt.start, 1, addr_opt.security};
+  for (uint64_t addr_index = addr_opt.start; addr_index < addr_opt.total; addr_index++) {
```
for getting the unused address `addr_opt.total` could be 0.

oopsmonk commented on cclient/api/extended/get_inputs.c:
```diff
@@ -13,51 +13,60 @@
+      ret_code = RC_OOM;
+      log_error(client_extended_logger_id, "%s create get balances request or response object failed: %s\n", __func__,
+                error_2_string(ret_code));
+      return ret_code;
```
memory leak

oopsmonk commented on cclient/api/extended/get_inputs.c:
```diff
@@ -13,51 +13,60 @@
+      return ret_code;
+    }
+
+    ret_code = iota_client_get_new_address(serv, seed, addr_opt_elt, &balances_req->addresses);
```
how about using `iota_sign_address_gen_flex_trits` instead of `iota_client_get_new_address`?

howjmay replied:
In this new code, I used it.

oopsmonk commented on cclient/api/extended/get_inputs.c:
```diff
@@ -4,63 +4,130 @@
 
-retcode_t iota_client_get_inputs(iota_client_service_t const* const serv, flex_trit_t const* const seed,
-                                 address_opt_t const addr_opt, uint64_t const threshold, inputs_t* const out_inputs) {
+static retcode_t is_unused_address(iota_client_service_t const* const serv, flex_trit_t const* const addr,
```
please call `is_unused_address` in get_new_address.c

howjmay replied:
but `is_unused_address` is a static function in `get_new_address.c` which is not able to be included by `get_inputs.c`
That's why I use this ugly solution

oopsmonk replied:
feel free to make `is_unused_address` public in get_new_address.c

oopsmonk commented on cclient/api/extended/get_inputs.c:
```diff
@@ -4,63 +4,130 @@
+  uint64_t addr_index, addr_index_end;
+  if (addr_opt.total == 0) {
+    addr_index = 0;
+    addr_index_end = (unsigned)-1;
```
`addr_index_end` is `uint64_t`

howjmay replied:
I want to assign maximun possible value of `uint64_t` to `addr_index_end`. Use an `(unsigned)-1` would be a simple way without including other header file

oopsmonk replied:
the maximum value of `uint64_t`(18446744073709551615) is no equal to `unsigned int -1`(4294967295)

oopsmonk commented on cclient/api/extended/get_inputs.c:
```diff
@@ -4,63 +4,130 @@
+    balances_req->threshold = 100;  // currently `100` should be used
+
+    ret_code = iota_client_get_balances(serv, balances_req, balances_res);
+    if (ret_code == RC_OK && addr) {
```
`addr` no need to check again.

oopsmonk commented on cclient/api/extended/get_inputs.c:
```diff
@@ -4,63 +4,130 @@
-      log_warning(client_extended_logger_id, "insufficient balance\n");
     }
+
+    get_balances_req_free(&balances_req);
```
you can reuse the request and response instead of allocating every time.

howjmay replied:
`hash243_queue_to_json_array` would convert all the elements in `hash243_queue_t` to json array,  so if I didn't free the `hash243_queue_t` variable, then `hash243_queue_to_json_array` would generate a wrong json array.
In the other hand, if reuse them without freeing the `get_balances_req_t` and `get_balances_res_t` object, then there will be memory leak problems, since the elements such as `hash243_queue_t` should be freed.

oopsmonk replied:
you can clean the queue instead of freeing the entire struct, isn't it?

howjmay replied:
sure

oopsmonk commented on cclient/api/extended/get_inputs.c:
```diff
@@ -4,63 +4,130 @@
   }
 
 done:
+  free(addr);
```
memory leak.

howjmay replied:
sorry I used `asan` and examine the code but I don't know which object cause memory leak.

oopsmonk replied:
how did you examine this API call? the `addr` has a memory leak issue from the logic, but I can update you a sanitizer report later.

oopsmonk replied:
```
==30356==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 324 byte(s) in 4 object(s) allocated from:
	#0 0x7f6e02487108 in calloc (/usr/lib/x86_64-linux-gnu/libasan.so.5+0xee108)
	#1 0x55d92ab43984 in iota_sign_address_gen_flex_trits common/helpers/sign.c:92
	#2 0x55d92a9fb86f in iota_client_get_inputs cclient/api/extended/get_inputs.c:76
	#3 0x55d92a9e9738 in example_get_inputs cclient/api/examples/example_get_inputs.c:25
	#4 0x55d92a9e4de7 in main cclient/api/examples/cclient_examples.c:88
	#5 0x7f6e01683b96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)
```

howjmay replied:
oh I got the reason why I didn't get the memory leak issue during testing last time!

howjmay commented:
Do we need to consider the situation that a user want to move all the balances to the unused address?

oopsmonk commented:
> Do we need to consider the situation that a user want to move all the balances to the unused address?

will that be a problem? I don't know what you were referring to?

howjmay commented:
Since if the user want to transfer all the balances into the unused address, then some steps of the current implementation are unnecessary, and we can call both `iota_client_get_new_address` and `iota_client_get_balances` once only to get all the balances we need which can save some time.

oopsmonk commented:
> Since if the user want to transfer all the balances into the unused address, then some steps of the current implementation are unnecessary, and we can call both `iota_client_get_new_address` and `iota_client_get_balances` once only to get all the balances we need which can save some time.

in general, this API helps to get enough balance from the account(seed), moving all balances could be done in other ways as you mentioned.

howjmay commented:
> > Since if the user want to transfer all the balances into the unused address, then some steps of the current implementation are unnecessary, and we can call both `iota_client_get_new_address` and `iota_client_get_balances` once only to get all the balances we need which can save some time.
> 
> in general, this API helps to get enough balance from the account(seed), moving all balances could be done in other ways as you mentioned.

Got it. Thank you for clarifying my worries.

Commit:
Merge branch 'develop' into iota_client_get_inputs_enhancement
